File: command/command/commands/InsertItem.cs
using command.document;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace command.commands
{
	public class InsertItem : AbstractCommand
	{
		private readonly IDocument _document;
		private readonly IDocumentItem _item;
		private readonly int _position;

		public InsertItem(IDocument document, IDocumentItem item, int position)
		{
			_document = document ?? throw new ArgumentNullException(nameof(document));
			_item = item ?? throw new ArgumentNullException(nameof(item));
			if (position > _document.ItemsCount)
			{
				throw new CommandError($"Invalid position {position}");
			}
			_position = position;
		}

		protected override void ExecuteImpl() => _document.InsertItem(_item, _position);
		protected override void UnexecuteImpl() => _document.DeleteItem(_position);

		protected override void DisposeImpl()
		{
			if (!IsExecuted)
			{
				_item.Dispose();
			}
		}
	}
}

File: command_java/Document/src/document/commands/InsertItem.java
package document.commands;

import document.documentItem.ItemsCollection.IDocumentItem;
import document.documentItem.ItemsCollection.IMutableDocumentItemCollection;
import hsitory.AbstractCommand;

import java.util.Optional;

class InsertItem extends AbstractCommand {

    public InsertItem(IMutableDocumentItemCollection items, IDocumentItem newItem, Optional<Integer> position) {
        _newItem = newItem;
        _items = items;
        _position = position.isPresent() ? position.get() : items.GetSize();
    }

    @Override
    protected void DestroyImpl() throws Exception {
        if (!IsExecuted())
        {
            _newItem.close();
        }
    }

    @Override
    protected void UnExecuteImpl() {
        _items.DeleteItem(_position);
    }

    @Override
    protected void ExecuteImpl() {
        _items.InsertItem(_newItem, _position);
    }

    private final IDocumentItem _newItem;
    private final IMutableDocumentItemCollection _items;
    private final int _position;
}
